City-of-Helsinki/drupal-helfi-rekry
# UHF-6346: Job listing migration fixes

# [UHF-6346](https://helsinkisolutionoffice.atlassian.net/browse/UHF-6346)
Made changes to the job listing migration.

## What was done
* Take into account translated nodes when unpublishing job listings that are missing from source.
* When importing job listings, skip importing _publish on_ date when the date is in the past and the corresponding node is already published. There is no need to import the date field, as the node state already represents a state after published by scheduler. This change prevents the situation when a past value is imported and the scheduler temporarily updates the node as unpublished.
* Small fix to error: Call to a member function hasField() on null.
* Refactored unpublishing job listings that are missing from source. Now reacting to MigrateMissingSourceRowsEvent so there is no longer e.g. a need to mark all previously imported nodes as ready to be re-imported.

## How to install
* Make sure your instance is up and running on correct branch.
  * `git pull origin UHF-6346-scheduler-publish-changes`
  * `make fresh`
* Run `make drush-cr`

## How to test
* Import job listings for the different languages (using `migrate:import`):
  * `helfi_rekry_jobs:all`
  * `helfi_rekry_jobs:all_sv`
  * `helfi_rekry_jobs:all_en`
* After importing, run cron so that the scheduler publishes and unpublishes nodes.
  * `drush cron`
* Go to the `/admin/content?title=&type=job_listing` page and check the job listings.
* Open the job listing source APIs to your browser.
* Compare the node publish status with the API. Should the nodes be published or not when comparing to the API publicationStarts and publicationEnds values?
  * Also check the different translations.
* Because you run the `make fresh` command, you should have some job listings that are missing from source API. Are those unpublised as they should be?
  * Try publishing some and the run the migrate:import again. Now they should be unpublished again.
* Check that code looks ok.

## Designers review
